include/engine/data_watchdog.hpp

```diff
@@ -92,22 +91,27 @@ class DataWatchdog
         // in that case we don't modify anthing
         if (shared_timestamp->timestamp == current_timestamp.timestamp)
         {
-            BOOST_ASSERT(shared_timestamp->layout == current_timestamp.layout);
-            BOOST_ASSERT(shared_timestamp->data == current_timestamp.data);
+            if (auto facade = cached_facade.lock())
+            {
+                BOOST_ASSERT(shared_timestamp->layout == current_timestamp.layout);
+                BOOST_ASSERT(shared_timestamp->data == current_timestamp.data);
 
-            return get_locked_facade();
+                return get_locked_facade(facade);
+            }
         }
 
         // this thread has won and can update the data
         boost::upgrade_to_unique_lock<boost::upgrade_mutex> unique_facade_lock(facade_lock);
 
         current_timestamp = *shared_timestamp;
```

Comment:
Should be check `if (shared_timestamp->timestamp == current_timestamp.timestamp)` also added here before `current_timestamp` assignment? This will prevent situation when two threads will try acquire a unique lock, and the second will reload facade with the same data